### ISSUE TITLE
input label false positive inside an iframe

### ISSUE THREAD

**Issue by damien-git** [dequelabs/axe-core#811]
I am getting a "Ensures every form element has a label" false positive with the following files. I tried to reduce that test to a minimum, but using an input id and label for attribute does not change the result. I was using the latest release.

file1.html:
```
<!DOCTYPE html>
<html lang="en">
<head>
    <title>Bug demo file 1</title>
    <script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
    <script src="axe.min.js"></script>
</head>
<body>
    <iframe id="iframe" src="file2.html"></iframe>
    <script>
        var iframeLoaded = function(e) {
            let iframe = document.getElementById('iframe');
            let iframeDoc = iframe.contentWindow.document;
            let body = iframeDoc.querySelector('body');
            let context = jQuery(body);
            axe.run(context, function(err, results) {
                if (err) throw err;
                console.log(results);
            });
        }
        iframe.addEventListener('load', iframeLoaded, false);
    </script>
</body>
</html>
```

file2.html:
```
<!DOCTYPE html>
<html lang="en">
<head>
    <title>Bug demo file 2</title>
</head>
<body>
    <form>
        <label>test<input></label>
    </form>
</body>
</html>
```

**Comment by WilcoFiers** [comment]
This example doesn't fail the label rule for me. I also don't see how that could even be possible unless there's something you've overlooked.

**Comment by damien-git** [comment]
I get a violation every time, using Firefox 52 ESR or Chromium 62. No other file is needed, just put these two files on a web server and open the first one.

**Comment by damien-git** [comment]
Make sure to name the second file `file2.html` (I forgot the extension in my initial post).

**Comment by dylanb** [comment]
This is invalid use of axe-core. You can simply include axe-core in file2.html thusly:

```
<!DOCTYPE html>
<html lang="en">
<head>
    <script src="point to the axe.min.js file"></script>
    <title>Bug demo file 2</title>
</head>
<body>
    <form>
        <label>test<input></label>
    </form>
</body>
</html>
```

and then call axe.run on the outer document.

**Comment by damien-git** [comment]
> This is invalid use of axe-core.

This is unfortunate. Since there is a mechanism to select elements inside iframes, I thought aXe was supporting validation inside an iframe.

Let me explain how I use it: the [Mura Accessibility plugin](https://github.com/MSU-NatSci/MuraAccessibility) is validating a whole site by loading pages in an iframe to validate them. The iframe is using `sandbox="allow-same-origin"` to prevent script execution inside and load pages faster, so I would not be able to use the aXe script in the iframe document.

It works 99% of the time: so far the input label false positive is the only issue we have noticed.

**Comment by dylanb** [comment]
@damien-git Yes, axe-core has iframe support built in - It is designed to work the way I described. 

axe-core uses many different APIs - which depend on things like the window and the document context. There is no guarantee that these API calls will work across iframe boundaries. In fact, I can pretty much guarantee that IDREF lookups will fail.

You could try the axe-core 3.0 library as this has more robust lookups - but once again - if you are not injecting correctly into sub-frames, your results will be incorrect.

**Comment by damien-git** [comment]
OK, thanks for your help.
